### PR TITLE
voting - first implementation 

### PR DESCRIPTION
**Pull request by tzachihakmon**

### CODE REVIEW

File: ServerSide/yoty/Core/Managers/Bids/BidsManager.cs [hunk 1]
```diff
@@ -402,5 +402,21 @@ private IEnumerable<BidEntity> GetSortBids(IEnumerable<BidEntity> bids, BidsSort
...
 
+        public async Task<Response> VoteForSupplier(VotingRequest votingRequest)
+        {
+            BidEntity bid = await _context.Bids.Where(b => b.Id == votingRequest.BidId).Include(b => b.CurrentProposals).FirstOrDefaultAsync().ConfigureAwait(false);
+            if (bid == null)
+            {
+                return new Response() { IsOperationSucceeded = false, SuccessOrFailureMessage = BidNotFoundFailString };
+            }
+
+            IEnumerable<SupplierProposalEntity> propsalsToUpdate = bid.CurrentProposals.Where(propsal => votingRequest.VotedSuppliersIds.Contains(propsal.SupplierId));
```
Comment (yotamc4) [inline ServerSide/yoty/Core/Managers/Bids/BidsManager.cs]: there will be at most one proposal here, so you can query with FirstOrDefaultAsync and get the Entity obj

File: ServerSide/yoty/Data/Entities/SupplierProposalEntity.cs [hunk 2]
```diff
@@ -14,6 +14,8 @@ public class SupplierProposalEntity
 
         public string SupplierName { get; set; }
 
+        public int Votes { get; set; }
```
Comment (yotamc4) [inline ServerSide/yoty/Data/Entities/SupplierProposalEntity.cs:17]: we should add this to the DTO also

File: ServerSide/yoty/WebApi/Controllers/BidsController.cs [hunk 3]
```diff
@@ -163,10 +163,10 @@ public async Task<ActionResult> DeleteBuyer(string bidId, string buyerId)
 
         [HttpDelete]
         [Route("{bidId}/proposals/{proposalId}")]
```
Comment (yotamc4) [inline ServerSide/yoty/WebApi/Controllers/BidsController.cs]: change the Routh to supplierId

File: ServerSide/yoty/WebApi/PublicDataSchemas/ClientRequest/VotingRequest.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    using Microsoft.AspNetCore.Mvc;
+    using Newtonsoft.Json;
+
+    public class VotingRequest
+    {
+        [JsonProperty("buyerId")]
+        public string BuyerId { get; set; }
+
+        [JsonProperty("chosenSuppliersIds")]
+        public string[] VotedSuppliersIds { get; set; }
```
Comment (yotamc4) [inline ServerSide/yoty/WebApi/PublicDataSchemas/ClientRequest/VotingRequest.cs]: Why list?

File: ServerSide/yoty/WebApi/Controllers/BidsController.cs [hunk 5]
```diff
@@ -176,6 +176,29 @@ public async Task<ActionResult> DeleteSupplierProposal(string bidId, string prop
             return this.StatusCode(StatusCodes.Status405MethodNotAllowed, response.SuccessOrFailureMessage);
         }
 
+        [HttpPost]
+        [Route("{bidId}/vote")]
+        public async Task<ActionResult> VoteForSupplier(VotingRequest votingRequest)
+        {
+            if (votingRequest.BidId == null || votingRequest.BuyerId == null)
```
Comment (yotamc4) [inline ServerSide/yoty/WebApi/Controllers/BidsController.cs]: Add a TODO / Add to this PR:
limit participant to only vote for a single proposal

File: ServerSide/yoty/Core/Managers/Bids/BidsManager.cs (hunk 1 above)
Comment (tzachihakmon) [reply ServerSide/yoty/Core/Managers/Bids/BidsManager.cs]: why only one ? I implemented it such that we are getting an array of supplier ids , so there could be more than one proposal ( proposal for supplier x and propals for supplier y , both are part of the ids array )

File: ServerSide/yoty/Core/Managers/Bids/BidsManager.cs [hunk 6]
```diff
@@ -404,18 +405,31 @@ private IEnumerable<BidEntity> GetSortBids(IEnumerable<BidEntity> bids, BidsSort
...
+
+            SupplierProposalEntity proposal = bid.CurrentProposals.Where(p => p.SupplierId == votingRequest.VotedSupplierId).FirstOrDefault();
+            if (proposal == null)
             {
-                proposal.Votes += 1;
+                return new Response() { IsOperationSucceeded = false, SuccessOrFailureMessage = $"no proposal for supplier: {votingRequest.VotedSupplierId} has found for the bid." };
             }
 
+            participancy.HasVoted = true;
+            proposal.Votes += 1;
```
Comment (yotamc4) [inline ServerSide/yoty/Core/Managers/Bids/BidsManager.cs:432]: The DB needs to be updated @tzachihakmon 
Add:

try
            {
                _context.Set<ParticipancyEntity>().Update(participancy)
                _context.Set<SupplierProposalEntity>().Update(proposal)
                await _context.SaveChangesAsync().ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                return new Response() { IsOperationSucceeded = false, SuccessOrFailureMessage = ex.Message };
            }

**Pull request merged**